src/infill/SierpinskiFill.cpp

```diff
@@ -0,0 +1,802 @@
+                first_dir = SierpinskiTriangle::SierpinskiDirection::AB_TO_BC;
+                second_dir = SierpinskiTriangle::SierpinskiDirection::AC_TO_BC;
+                break;
+            case SierpinskiTriangle::SierpinskiDirection::AC_TO_BC:
+                first_dir = SierpinskiTriangle::SierpinskiDirection::AB_TO_BC;
+                second_dir = SierpinskiTriangle::SierpinskiDirection::AC_TO_AB;
+                break;
+        }
+        sub_root.children.emplace_back(middle, t.a, t.straight_corner, first_dir, !t.straight_corner_is_left, t.depth + 1);
+        sub_root.children.emplace_back(middle, t.straight_corner, t.b, second_dir, !t.straight_corner_is_left, t.depth + 1);
+        for (SierpinskiTriangle& child : t.children)
+        {
+            createTree(child);
+        }
+    }
+}
+void SierpinskiFill::createTreeStatistics(SierpinskiTriangle& triangle)
+{
+    Point ac = triangle.straight_corner - triangle.a;
+    float area = 0.5 * INT2MM2(vSize2(ac));
+    float short_length = .5 * vSizeMM(ac);
+    float long_length = .5 * vSizeMM(triangle.b - triangle.a);
+    triangle.area = area;
+    triangle.realized_length = (triangle.dir == SierpinskiTriangle::SierpinskiDirection::AC_TO_BC)? long_length : short_length;
+    for (SierpinskiTriangle& child : triangle.children)
+    {
+        createTreeStatistics(child);
+    }
+}
+
+
+void SierpinskiFill::createTreeRequestedLengths(SierpinskiTriangle& triangle)
+{
+    if (triangle.children.empty())
+    { // set requested_length of leaves
+        float density = density_provider(triangle.a, triangle.b, triangle.straight_corner, triangle.straight_corner);
+        triangle.requested_length = density * triangle.area / INT2MM(line_width);
+    }
+    else
+    { // bubble total up requested_length and total_child_realized_length
+        for (SierpinskiTriangle& child : triangle.children)
+        {
+            createTreeRequestedLengths(child);
+            
+            triangle.requested_length += child.requested_length;
+            triangle.total_child_realized_length += child.realized_length;
+        }
+    }
+}
+
+void SierpinskiFill::createLowerBoundSequence()
+{
+    sequence.emplace_front(&root);
+
+    if (deep_debug_checking) debugCheck();
+    for (int iteration = 0; iteration < 999; iteration++)
+    {
+        bool change = false;
+        change |= subdivideAll();
+        if (deep_debug_checking) debugCheck();
+        
+        if (constraint_error_diffusion)
+        {
+            change |= bubbleUpConstraintErrors();
+            if (deep_debug_checking) debugCheck();
+        }
+        
+        
+        if (!change)
+        {
+            std::cerr << "Finished after "<<(iteration + 1)<<" iterations, with max_depth = "<<max_depth << '\n';
+            break;
+        }
+    }
+}
+
+bool SierpinskiFill::subdivideAll()
+{
+    std::vector<std::list<std::list<SierpinskiTriangle*>::iterator>> depth_ordered;
```

Comment:
More performant? Why do you think that?
Maybe something with cache misses or whatever.

Changed it in 631ec56aee7714ecd37ba8191ff1dc94e3735e40

Reply:
Yes, a bit like cache misses. Vector is a "contiguous data structure", meaning that all elements of the vector are listed after one another in memory. When a computer has to look up an element in RAM, it requests a whole block of memory at a time, which for small elements like pointers would read like 8 pointers at a time. So in a vector it would read the first pointer slowly and then the next 7 pointers are already in the chip's L1 cache so those are almost instantaneous.

In `std::list`, it has to look up the location of the next list element every time, which may be in a completely different location in RAM. If you're iterating over an `std::list` it will take an order of magnitude longer.